synapse/storage/database.py

```diff
@@ -796,6 +800,20 @@ async def runInteraction(
             The result of func
         """
 
+        async def _run_callbacks(callbacks: List[_CallbackListEntry]):
+            sync_callbacks: List[_CallbackListEntry] = []
+
+            for cb, args, kwargs in callbacks:
+                if inspect.iscoroutinefunction(cb):
+                    awaitable = cb(*args, **kwargs)
+                    assert isinstance(awaitable, Awaitable)
+                    await awaitable
+                else:
+                    sync_callbacks.append((cb, args, kwargs))
+
+            for cb, args, kwargs in sync_callbacks:
+                cb(*args, **kwargs)
```

Comment:
can you add some comments here please? What is going on here, and why?

Reply:
Added a docstring to describe it - should this be hoisted to a class method now perhaps? https://github.com/matrix-org/synapse/pull/13308/commits/b8ea2136ec4296c16c24d757bca34d74b9e76dd6

Reply:
> should this be hoisted to a class method now perhaps?

or, better, a top-level function.